src/Common/ZooKeeper/ZooKeeperImpl.cpp

```diff
@@ -1278,6 +1276,22 @@ void ZooKeeper::receiveEvent()
             response->removeRootPath(root_path);
         }
 
+        /// Find from operations
+        if (request_info.watch)
+        {
+            bool add_watch = false;
+            if (request_info.isExists)
```

Comment:
The idea is Ok.

Maybe dynamic_cast will be better than adding a field?

Reply:
Also please add some explanation in comments.

Reply:
getOpNum may also suffice.

Reply:
I changed to use the getOpNum. Seems the test case `test_chroot_with_same_root` failed. It should be not related to this change, I need to look at why.

Reply:
There were no failures of this test before.

Reply:
Yes, this error occurs after this PR. I fixed your comment first, I mean this test failure should still have, it is caused by other place in this PR.